function_and_scope_questions.js

```diff
@@ -0,0 +1,199 @@
+    };
+  }
+  function times(x) {
+    return function (y) {
+      return x * y;
+    };
+  }
+  function dividedBy(x) {
+    return function (y) {
+      return Math.floor(y / x);
+    };
+  }
+  
+
+  function runSolutionOne(){
+    let answer=seven(times(five()));
+    console.log(answer);
+  }
+
+  runSolutionOne();
+
+
+// ------------------------------------------------------------------------------------------------------
+
+
+
+//QUESTION 2:
+/* You are going to be given a word. Your job is to return the middle character of the word. If the word's length is odd,
+return the middle character.
+ If the word's length is even, return the middle 2 characters. */
+
+
+
+//SOLUTION:
+
+function getMiddlePartofInputString(inputString) {
+    //Code goes here!
+    
+    let halfLengthOfString=inputString.length / 2;
+    let lengthIsEven=inputString.length % 2 == 0;
+    
+    if (lengthIsEven) {
+        return inputString.substring((halfLengthOfString - 1), (halfLengthOfString + 1));
+    }
+    else {
+        return inputString.substring(Math.floor(halfLengthOfString), (Math.floor(halfLengthOfString) + 1));
+    }
+}
+
+function run_Solution_2() {
+    console.log(getMiddlePartofInputString('abcde')); // must return 'c'
+    console.log(getMiddlePartofInputString('efgh'));  // must return 'fg'
+}
+run_Solution_2();
+
+
+// --------------------------------------------------------------------------------------------------------------------------------
+
+
+
+
+
+
+
+// QUESTION 3:
+/* Write a function which partitions a list of items based on a given predicate.
+
+After the partition function is run, the list should be of the form [ F, F, F, T, T, T ]
+where the Fs (resp. Ts) are items for which the predicate function returned false (resp. true).
+
+NOTE: the partitioning should be stable; in other words:
+the ordering of the Fs (resp. Ts) should be preserved relative to each other.
+
+For convenience and utility, the partition function should return the boundary index.
+In other words: the index of the first T value in items. */
+
+
+// SOLUTION 3:
+
+
+
+// partition the items array so that all values for which pred returns true are
+// at the end, returning the index of the first true value
+
+function partitionOn(items) {
+
+
+    var newarray = [];
+    function firstFillAllOddNumbers(items) {
+        items.forEach(item => {
+            if (item % 2 !== 0) {
+                newarray.push(item);
+            }
+        });
+    }
+
+    function secondFillAllEvenNumbers(items) {
+        items.forEach(item => {
+            if (item % 2 == 0) {
+                newarray.push(item);
+            }
+        });
+    }
+    firstFillAllOddNumbers(items);
+    secondFillAllEvenNumbers(items);
+
+
+
+
+    var booleanArray = [];
+    newarray.forEach(element => {
+        booleanArray.push(element % 2 == 0);
+    }); // Filling boolean_Array with Boolean Values
+
+
+    var indexOfFirstTrueValue = booleanArray.indexOf(true);
```

Comment:
can we have some other name? This doesn't sound good;

Reply:
One of the two purposes of the program was to give the index of first true vale in the array, so only writing "index" or "firstTrueIndex" isn't fitting correctly. 
Which is why I had no other option but to write the complete "indexOfFirstTrueValue"